x-pack/plugins/code/server/routes/redirect_socket.ts

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Server } from 'hapi';
+import SocketIO, { Socket } from 'socket.io';
+import ClientIO from 'socket.io-client';
+
+import { SocketKind } from '../../model';
+import { Log } from '../log';
+import { ServerOptions } from '../server_options';
+import { BASE_PLACEHOLDER, mainNodeBaseUrl } from './redirect';
+
+export async function redirectSocketRoute(server: Server, serverOptions: ServerOptions, log: Log) {
+  const socketIO = SocketIO(server.listener, { path: '/ws' });
+  let redirectUrl = serverOptions.redirectToNode;
+  const hasBaseUrl = redirectUrl.includes(BASE_PLACEHOLDER);
+
+  async function connectToMainNode() {
+    let baseUrl = '';
+    if (hasBaseUrl) {
+      baseUrl = await mainNodeBaseUrl(redirectUrl);
+      redirectUrl = serverOptions.redirectToNode.replace(BASE_PLACEHOLDER, '');
+    }
+    const socketUrl = `${redirectUrl}`;
+    return ClientIO(socketUrl, { path: `${baseUrl}/ws` });
```

Comment:
This connection does not work if the `redirectUrl` points to an instance of Kibana secured with `https`, where the server's certificate is not inherently trusted by the node runtime. You'll need to respect/trust the certificate authorities specified at `server.ssl.certificateAuthorities` in order to establish this connection in this scenario.

Reply:
In my testing, I wasn't able to use the Code app at all when connecting to a non-code node. The repository list did not populate, and I couldn't create any new repositories either.

I tried to test this again just now, but I can't get any of the code routes to work on even a single node instance without this configuration:
![image](https://user-images.githubusercontent.com/3493255/51539321-ae632d00-1e21-11e9-87cb-e31f81d077c9.png)

Reply:
Ah, sorry. There is a recent change which requires you to add `xpack.code.codeNode: true` into the `kibana.yml` to enable the code node. I know the documentation has not caught up yet. :)

> It's not a security issue, but it does appear to render the multi-instance setup unusable when custom CAs are specified.

Yes, we haven't thought about this case when implementing the current multi-instance setup. Just filed an issue in here: https://github.com/elastic/code/issues/836. This is definitely a bug we need to fix before the beta going out of the door.

Reply:
Thanks!